docs/development/sync-protocol.md

```diff
@@ -0,0 +1,431 @@
+
+```mermaid
+sequenceDiagram
+   client ->> server: POST /preflight/<machine_id>
+   server -->> client: preflight response (200)
+   client ->> server: POST /eventupload/<machine_id>
+   server -->> client: eventupload response (200)
+   client ->> server: GET /ruledownload/<machine_id>
+   server --> client: ruledownload response (200)
+   client ->> server: POST /postflight/<machine_id> request
+   server -->> client: postflight response (200)
+```
+
+Where `<machine_id>` is a unique string identifier for the client by default
+Santa makes this is the string `IOPlatformUUID` from IOKit's
+`IOPlatformExpertDevice` aka the hardware UUID.
+
+## Authentication
+
+The protocol expects to the client to authenticate the server via SSL/TLS. Additionally Sync Server's may support client certificates and use mutual TLS.
+
+## Stages
+
+All URLs are of the form `/<stage_name>/<machine_id>`, e.g. the preflight URL is `/preflight/<machine_id>`.
+
+### Preflight
+
+The preflight stage is used by the client to report host information to the Sync
+Server and to retrieve a limited set of configuration settings from the server.
+These configuration options override the initial values set from the application
+configuration profile.
+
+This follows the following transaction:
+
+```mermaid
+sequenceDiagram
+   client ->> server: POST /preflight/<machine_id>
+   server -->> client: preflight response
+```
+
+#### `preflight` Request
+The request consists of the following JSON keys:
+
+| Key | Required | Type | Meaning | Example Value | 
+|---|---|---|---|---|
+| serial_num    | YES | string | The macOS serial number from IOKit `kIOPlatformSerialNumberKey` |  "XXXZ30URLVDQ" |
+| hostname      | YES | string | The FQDN hostname of the client | markowsky.example.com | 
+| os_version    | YES | string | The OS version of the client from /System/Library/CoreServices/SystemVersion.plist | 12.4 |
+| os_build      | YES | string | The OS build from /System/Library/CoreServices/SystemVersion.plist | "21F5048e" |
+| model_identifier | NO | string | The model of the macOS system  | | 
+| santa_version | YES | string | 2022.3 |
+| primary_user  | YES | string | The username | markowsky |
+| fcm_token     | NO | string | Firebase cloud messaging token (for push notifications) |  |
+| binary_rule_count | NO | int | Number of binary allow / deny rules the client has at time of sync| 1000 |
+| certificate_rule_count | NO | int | Number of certificate allow / deny rules the client has at time of sync | 3400 |
+| compiler_rule_count | NO | int | Number of compiler rules the client has time of sync |
+| transitive_rule_count | NO | int | Number of transitive rules the client has at the time of sync |
+| teamid_rule_count | NO | int | Number of TeamID rules the client has at the time of sync | 24 |
+| client_mode | YES | string | the mode the client is operating in, either "LOCKDOWN" or "MONITOR" | LOCKDOWN |
+| clean_sync | NO | bool | true |
```

Comment:
Add a meaning for this - it's useful to explain that a clean sync can be requested by both client and server